lgc/state/PalMetadata.cpp

```diff
@@ -240,11 +295,89 @@ void PalMetadata::setUserDataSpillUsage(unsigned dwordOffset) {
     *m_spillThreshold = m_document->getNode(dwordOffset);
 }
 
+// =====================================================================================================================
+// Fix up user data registers. Any user data register that has one of the unlinked UserDataMapping values defined
+// in AbiUnlinked.h is fixed up by looking at pipeline state.
+void PalMetadata::fixUpRegisters() {
+  static const unsigned ComputeRegRanges[] = {mmCOMPUTE_USER_DATA_0, 16};
+  static const unsigned Gfx8RegRanges[] = {mmSPI_SHADER_USER_DATA_PS_0, 16, mmSPI_SHADER_USER_DATA_VS_0, 16,
+                                           mmSPI_SHADER_USER_DATA_GS_0, 16, mmSPI_SHADER_USER_DATA_ES_0, 16,
+                                           mmSPI_SHADER_USER_DATA_HS_0, 16, mmSPI_SHADER_USER_DATA_LS_0, 16};
+  static const unsigned Gfx9RegRanges[] = {mmSPI_SHADER_USER_DATA_PS_0, 32, mmSPI_SHADER_USER_DATA_VS_0, 32,
+                                           mmSPI_SHADER_USER_DATA_ES_0, 32, mmSPI_SHADER_USER_DATA_HS_0, 32};
+  static const unsigned Gfx10RegRanges[] = {mmSPI_SHADER_USER_DATA_PS_0, 32, mmSPI_SHADER_USER_DATA_VS_0, 32,
```

Comment:
Maybe arrays of `std::pair<unsigned, unsigned>` with nested-braces for initialization? A quick check suggests that that also gets emitted as .rodata.